### ISSUE TITLE
Invalid rollover of random portion of ULID

### ISSUE THREAD

**Issue by grempe** [ryan-mars/ulid-workers#8]
From comment made by @peterbourgon 

https://github.com/ryan-mars/ulid-workers/pull/7#issuecomment-1035370091

It looks like if you give incrementBase32 a ULID with the entropy section maxed out, it will roll the entropy over to zero, and increment the timestamp by 1ms.

```
$ ulid 01FVJFV41BZZZZZZZZZZZZZZZZ
Thu Feb 10 18:55:29.835 UTC 2022

$ incrementBase32('01FVJFV41BZZZZZZZZZZZZZZZZ')
01FVJFV41C0000000000000000

$ ulid 01FVJFV41C0000000000000000
Thu Feb 10 18:55:29.836 UTC 2022
```

Oops!

**Comment by grempe** [comment]
Good find @peterbourgon!

I'll add a test and a fix. Perhaps you'd also like to file this bug upstream where it originated prior to this fork?

I will note though that I think you'd need to monotonically generate `2^80` ULIDs within the same millisecond in order to exercise this rollover. Right? Or you'd have to be very very very unlucky with the initial output of `getRandomValues()` while generating monotonically.

Just for fun, I believe it would take ~38,000 years just to generate `2^80` ULIDs monotonically (assuming you can make a trillion a second). Doing so within the same millisecond is clearly outside the reach of humankind.

https://www.wolframalpha.com/input?i=1208925819614629174706176+%2F+1000000000000+%2F60+%2F60+%2F24%2F365

**Comment by peterbourgon** [comment]
> I will note though that I think you'd need to monotonically generate 2^80 ULIDs within the same millisecond in order to exercise this rollover. Right?

My understanding is that the first ULID generated by the factory for a given millisecond has a random value between `0000000000000000` and `ZZZZZZZZZZZZZZZZ` as its entropy component, and that subsequent ULIDs generated in the same millisecond increment that initial random value by 1. If that's true, then you don't need to generate 2^80 IDs to hit the issue, you just need your initial random value to be near enough to the maximum value.

> the time component of the ULID is never passed to incrementBase32(), only the random component of the ULID is ever passed.

Ah, I see. And it appears that, prior to c74de1c, if that random component was `ZZZZZZZZZZZZZZZZ` the function returns `Error: Incorrectly encoded string`. Although I will observe that ULID generation failing in this way would likely be somewhat surprising to users :) Of course, ULID generation with monotonic entropy can always potentially fail! That's unavoidable. You just want to make sure the stakes are clear to your users.

**Comment by grempe** [comment]
> My understanding is that the first ULID generated by the factory for a given millisecond has a random value between 0000000000000000 and ZZZZZZZZZZZZZZZZ as its entropy component, and that subsequent ULIDs generated in the same millisecond increment that initial random value by 1. If that's true, then you don't need to generate 2^80 IDs to hit the issue, you just need your initial random value to be near enough to the maximum value.

Yes, that's my understanding too.  That's why I said:

> you'd have to be very very very unlucky with the initial output of getRandomValues() while generating monotonically

:-)

I think in practical terms you'd have to hit the all `Z` lottery in the millisecond you want to generate more than one ULID. Not impossible, just infinitesimally unlikely. In any case, with the enhancements I just added you would now receive a thrown exception to handle in your application code. No surprises I think.

**Comment by peterbourgon** [comment]
> you'd have to be very very very unlucky

I guess so, but I guess it's also better if a software package like this one is _actually_ sound, not just _probably_ sound :) Which it is now!

**Comment by ryan-mars** [comment]
I'm not sure how I feel about throwing an error when the user is trying to generate a ULID. It is unexpected behavior and the types and documentation don't make it clear that this is a possibility. 

We should either:
* make it clear that sequential calls to a monotonic factory could throw in rare situations where random overflows
* make it clear that if randomness overflows it's the users job to verify `a < b` for each ULID (not cool IMHO)
* make it clear that if randomness overflows seed will increment by 1ms (could break spec)

If we throw, the error message should be helpful.

**Comment by peterbourgon** [comment]
> which the normal use case never does as I explained above

This isn't true — it certainly can happen. It's statistically improbable, true, but that doesn't mean "never".

**Comment by grempe** [comment]
>> which the normal use case never does as I explained above

> This isn't true

As you can see from the code sample I referenced I was speaking in the context of the possibility of passing a string of length greater than 16 into the increment function.

Which is not possible to do in normal public import operation of the library. I stand by the statement and I don't really appreciate the accusatory tone @peterbourgon.

**Comment by peterbourgon** [comment]
Oh, we're speaking about different things. Mea culpa.

**Comment by ryan-mars** [comment]
> I stand by the statement and I don't really appreciate the accusatory tone 

@grempe I didn't read any tone into @peterbourgon's statements but either way it's best to assume everyone here has the best intentions. 

Furthermore I find @peterbourgon's  thought process helpful and I wouldn't want either of you two to lose any desire to contribute.

**Comment by peterbourgon** [comment]
> to avoid it

Can't really avoid it! Can only mitigate it. With non-monotonic entropy, each millisecond permits the full domain of 2^80 possible values. But opting in to monotonic entropy necessarily reduces that set of choices. In either case, when entropy is exhausted, generation must either fail, or produce duplicate values.

**Comment by ryan-mars** [comment]
`$ ulid 01FVJFV41BZZZZZZZZZZZZZZZZ
Thu Feb 10 18:55:29.835 UTC 2022

$ ulid 01FVJFV41C0000000000000000
Thu Feb 10 18:55:29.836 UTC 2022`

Wait a minute. This is still monotonic behavior. This should not throw. Given the context of a Cloudflare worker, where the date seed is unreasonably static would it be so bad to increment time by 1ms? If this were an environment where I have high resolution timer incrementing date seed would violate the intent of the caller. However in a Cloudflare worker incrementing the date seed would be harmless. 

We can explain that this is expected behavior, and a rare possibility. If the user cares that much about keeping the same time in an environment without high resolution timers they can verify it themselves. 

What would be undesirable in this context is a failed worker invocation because of an uncaught exception in an _id generator_. 

I'd like to update the docs and allow for random of Z... to roll over to the next seed.

**Comment by peterbourgon** [comment]
> would it be so bad to increment time by 1ms?

It's not correct to do so 🤷

**Comment by ryan-mars** [comment]
> when entropy is exhausted, generation must either fail, or produce duplicate values. 

Why "must"? Is that the spec or some other principle?

**Comment by ryan-mars** [comment]
> > would it be so bad to increment time by 1ms?
> 
> 
> 
> It's not correct to do so 🤷 

LOL says who?

**Comment by peterbourgon** [comment]
> Why "must"? Is that the spec or some other principle?

ULIDs are 48 bits of millisecond-precision time, plus 80 bits of entropy. In theory time isn't something you get to choose, it's an externality, implementations only have autonomy over the entropy part. It's computers so at the end of the day nothing really matters and you can do whatever you want, but nominally this means in each millisecond you get a theoretical maximum of 1.21e+24 IDs before the U part of the ULID fails. And monotonic entropy reduces that number inversely proportion to its predictability.

**Comment by grempe** [comment]
ok, let me try again.

THIS IS NOT A BUG.

## Summary

This bug report is fundamentally invalid with its main premise that there is a problem where an attempt to increment the max (`ZZZZZZZZZZZZZZZZ`) Base32 value will result in an inappropriate rollover, possibly even affecting the time component of the ULID.

This was the example provided.  We've established already that this example shown here will NEVER happen in any usage of the library. It is not possible. I will take pains to explain, again, why this is so.

False example:

```
$ ulid 01FVJFV41BZZZZZZZZZZZZZZZZ
Thu Feb 10 18:55:29.835 UTC 2022

$ ulid 01FVJFV41C0000000000000000
Thu Feb 10 18:55:29.836 UTC 2022
```

The reason this will never happen is because @peterbourgon was directly calling a private function `incrementBase32` with a full length ULID (time + random) when it was only expecting the last 16 chars of the ULID (random) for it to increment. This is improper use of a private function and not a bug. Full stop.

The `incrementBase32` function in question is only ever called once in the code. It NEVER has access to the time variable `lastTime`, but can only increment the `lastRandom` variable that it has access to.

https://github.com/ryan-mars/ulid-workers/blob/main/src/ulid.ts#L150

Based on these facts, this bug is invalid.

Even though the bug is invalid, I think there is an opportunity to slightly improve the experience for developers contributing to this library, and end users in one specific and extraordinarily rare case.

## Improvement One

I added a guard that checks for improper length (which throws an error) for calls to the `incrementBase32` function.  This private function and should never be called directly outside of a testing context. That is why this function is not available as a standard import and is "hidden" behind the `exportedForTesting` object that makes these functions available to the test suite. At no time should it be necessary for an end user to ever import or use this private function. This string length guard is helpful to prevent this improper usage of the function by developers, and is helpful in testing.

```typescript
    if (str.length > RANDOM_LEN) {
        throw new Error(`Base32 value to increment cannot be longer than ${RANDOM_LEN} characters`);
    }
```

https://github.com/ryan-mars/ulid-workers/blob/timefix/src/ulid.ts#L74

It could be pruned if necessary without breaking any functionality for the user.

## Improvement Two

The false premise of this bug report is that there can be a Base32 rollover that can improperly modify the time component of the ULID. This is NOT the case.

There is NO rollover bug in the code of this lib as currently coded (before or after my changes, so to be clear I have no cause to defend my own changes) that will modify the time component of the ULID when `incrementBase32` is called on a `ZZZZZZZZZZZZZZZZ` max value. Full stop.

>> It's not correct to do so 🤷
> LOL says who?

The ULID spec says so. To wit.

> Any attempt to decode or encode a ULID larger than this should be rejected by all implementations, to prevent overflow bugs.

https://github.com/ulid/spec#overflow-errors-when-parsing-base32-strings

The spec demands (and all prior Javascript implementations implement) that the lib will throw an error on attempted monotonic rollover of a maximum Base32 value.

The prior, and new, behavior is to throw an error if a user tries to encode or decode a larger value than the allowed max, OR if the call to `incrementBase32` would result in a rollover past `ZZZZZZZZZZZZZZZZ`. The only thing that has changed is the quality of the error message string.

Here is the pre-existing code that does so.

https://github.com/ryan-mars/ulid-workers/blob/main/src/ulid.ts#L76

Side note, there is also a second code path that would throw an error in that function that is also preexisting. I am not sure when/how this error can ever be reached (it would require the potential output to not be a string).

https://github.com/ryan-mars/ulid-workers/blob/main/src/ulid.ts#L87

The only thing that has changed for this potential improvement is the addition of the following guard early in the execution of this function:

```typescript
    if (str === "Z".repeat(RANDOM_LEN)) {
        throw new Error(`Cannot increment Base32 maximum value ${"Z".repeat(RANDOM_LEN)}`);
    }
```

https://github.com/ryan-mars/ulid-workers/blob/timefix/src/ulid.ts#L78

This is just a short circuit to accomplish the same goal as the already existing code implements but to do so as quickly as possible in the execution of the function. Namely, bail out if the increment would result in a rollover past `ZZZZZZZZZZZZZZZZ`. This is what the spec calls for.


## Additional comments

> would it be so bad to increment time by 1ms?

I do not agree with this approach.  Firstly, it is solving a problem that does not actually exist (the rollover of time) and that complexity is only preventing throwing an error in the infinitesimal chance of `getRandomValues()` providing a value that would cause a rollover within the 1ms window.

The spec had the forethought to address what should happen in the case of an overflow, and that is to exit.  Implementing this "fix" would actually introduce the concept of time to `incrementBase32` that it does not currently have and is inappropriate.

We should be acutely aware that it is unlikely that anyone would EVER reach this exception while even running this library continuously for thousands of years which some quick back of the napkin math shows. And even in this case, the worst thing that would happen is what the spec demands! That the program 'reject' an attempt to do so by throwing an error. If the user is concerned about this possibility they can wrap the call to this function in a `try/catch` with a retry which they should be doing anyway if they are that concerned. Most end users do not have this intolerance to extremely rare failures.

Some probably wrong, but directionally correct math:

https://github.com/ryan-mars/ulid-workers/issues/8#issuecomment-1035462000

For fun, and to allow your concerns that I am being dismissive of the probability of users seeing this exception, watch this video from the beginning up to about the 2:56 mark to get a sense for what 2^80 really represents.

https://www.youtube.com/watch?v=S9JGmA5_unY

In summary, we have all spent way too much of our day talking about a bug that does not actually exist, but which did highlight the utility of a couple of simple messaging additions that would help make the reason for failure more clear and which I went ahead and added in my branch.

Any further discussion of the merits of using monotonic code is not germane to the issue at hand as that decision has already been made. I would recommend closing this bug as not a bug, and we can discuss the error messaging text when I land the PR (or pull it out altogether, but I think it makes things clearer for developers reading the code).

**Comment by grempe** [comment]
Closing this issue I opened as the related PR implementing the improvements has been merged. If anyone wants to discuss more please re-open or let me know.
